enhancements/two-nodes-openshift/2no.md

```diff
@@ -0,0 +1,664 @@
+controlPlane:
+  name: master
+  replicas: 2
+metadata:
+  name: <cluster-name>
+platform:
+  none:
+    fencingCredentials:
+      bmc:
+          address: ipmi://<out_of_band_ip>
+          username: <user>
+          password: <password>
+pullSecret: ''
+sshKey: ''
+```
+
+For platform baremetal, a valid configuration is quite similar.
+```
+apiVersion: v1
+baseDomain: example.com
+compute:
+- name: worker
+  replicas: 0
+controlPlane:
+  name: master
+  replicas: 2
+metadata:
+  name: <cluster-name>
+platform:
+  baremetal:
+    fencingCredentials:
+      bmc:
+          address: ipmi://<out_of_band_ip>
+          username: <user>
+          password: <password>
+    apiVIPs:
+      - <api_ip>
+    ingressVIPs:
+      - <wildcard_ip>
+pullSecret: ''
+sshKey: ''
+```
+
+Unfortunately, Baremetal Operator already has a place to specify bmc credentials. However, providing credentials like this will result in conflicts as both the
+Baremetal Operator and the pacemaker fencing agent will have control over the machine state. In short, this example shows an invalid configuration that we must check for
+in the installer.
+```
+apiVersion: v1
+baseDomain: example.com
+compute:
+- name: worker
+  replicas: 0
+controlPlane:
+  name: master
+  replicas: 2
+metadata:
+  name: <cluster-name>
+platform:
+  baremetal:
+    fencingCredentials:
+      bmc:
+          address: ipmi://<out_of_band_ip>
+          username: <user>
+          password: <password>
+    apiVIPs:
+      - <api_ip>
+    ingressVIPs:
+      - <wildcard_ip>
+    hosts:
+      - name: openshift-master-0
+        role: master
+        bmc:
+          address: ipmi://<out_of_band_ip>
+          username: <user>
+          password: <password>
+      - name: <openshift_master_1>
+        role: master
+        bmc:
+          address: ipmi://<out_of_band_ip>
+          username: <user>
+          password: <password>
+pullSecret: ''
+sshKey: ''
+```
+
+### Topology Considerations
+
+2NO represents a new topology and is not appropriate for use with HyperShift, SNO, or MicroShift
+
+#### Standalone Clusters
+
+Two-node OpenShift is first and foremost a topology of OpenShift, so it should be able to run without any assumptions of a cluster manager. To achieve this, we will need to enable the installation
+of two-node clusters via the Agent-Based Installer to ensure that we are still meeting the installation requirement of using only 2 nodes.
+
+### Implementation Details/Notes/Constraints
+
+While the target installation requires exactly 2 nodes, this will be achieved by proving out the "bootstrap plus 2 nodes" flow in the core installer and then using assisted-service-based installers to bootstrap from one of the target machines to remove the requirement for a bootstrap node.
+
+So far, we've discovered topology-sensitive logic in ingress, authentication, CEO, and the cluster-control-plane-machineset-operator. We expect to find others once we introduce the new infrastructure topology.
```

Comment:
Is there discussion somewhere about the issues found and the mitigations? In paritcular I'm interested in the CPMSO parts. I suspect we want to disable the feature on this toplogy. It relies on Machine API anyway, are we expecting MAPI to be available as part of this toplogy?

Reply:
Machine API isn't relevant for this topology (in its initial run). If requirements come for compute nodes, then that could change.

I outlined the sensitivities I found in: https://docs.google.com/document/d/15NOPO50aLC5onUP-m6cVy1CguvCVfJmLYiKcREazbQE/edit?tab=t.0#bookmark=id.5y5cwxlcsndw

CPMSO can be disabled (we tested that early on). I don't fully understand what problems it's supposed to solve, so I assumed some follow up discussions with the maintainers before I declared a resolution path.

Reply:
CPMSO is intended to provide an automated way to update control plane machines, but if there's no machine API, there's no CPMSO. Have you considered the capabilities API and just disabling Machine API completely on this toplogy? And potentially other capabilities that don't correlate?

Reply:
I appreciate the suggestion about reviewing the capabilities API as a whole. I will add a task for that.

As far as disabling CPMSO, I will revisit how we disable that after reviewing the capabilities API. :)
It may be prudent to disable the CPMSO as well as the machine API in the first pass, so that there is still a level of protection if something down the line pushes us to enable machine API.